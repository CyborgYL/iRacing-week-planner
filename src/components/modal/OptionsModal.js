// @flow

import * as React from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';
import toggleIdInCollection from '../../lib/toggleIdInCollection';
import Modal from './Modal';
import Checkbox from '../Checkbox';
import availableColumns from '../../data/availableColumns';

import styles from '../../styles/main.module.scss';

type Props = {
  onClose: () => void,
  isOpen?: boolean,
  columnIds?: Array<string>,
  saveOptions: (string, Array<string>) => void,
};

export default function OptionsModal({ onClose, isOpen, columnIds, saveOptions }: Props) {
  const getColumnToggler = (id) => (newValue) => {
    const newColumns = toggleIdInCollection(columnIds, id, newValue);
    saveOptions('columns', newColumns);
  };

  const { t } = useTranslation();

  return (
    <Modal isOpen={isOpen} onClose={onClose} title={t('Options')} doneAction={onClose}>
      <div className={styles['container-fluid']}>
        <h5>{t('Columns')}</h5>
        <div className={styles.row}>
          {availableColumns.map((column) => (
            <div className={`${styles['col-xs-3']} ${styles['col-md-2']}`} key={column.id}>
              <Checkbox
                id={`options-columns-${column.id}`}
                disabled={column.forced === true}
                checked={columnIds.indexOf(column.id) !== -1}
                onChange={getColumnToggler(column.id)}
              >
                {t(column.header)}
              </Checkbox>
            </div>
          ))}
        </div>
      </div>
    </Modal>
  );
}

OptionsModal.defaultProps = {
  isOpen: false,
  columnIds: [],
};
