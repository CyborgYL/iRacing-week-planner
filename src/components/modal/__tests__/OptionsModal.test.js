import { describe, test } from '@jest/globals';
import React from 'react';
import renderer, { act } from 'react-test-renderer';
import OptionsModal from '../OptionsModal';

describe('components/modal/OptionsModal', () => {
  test('renders closed', async () => {
    const component = renderer.create(
      <OptionsModal isOpen={false} onClose={() => {}} columnIds={[]} saveOptions={() => {}} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('renders with checkbox saving', async () => {
    const save = jest.fn(() => {});
    let component;
    act(() => {
      component = renderer.create(
        <OptionsModal isOpen onClose={() => {}} columnIds={['id', 'class']} saveOptions={save} />,
      );
    });

    expect(component.toJSON()).toMatchSnapshot();

    await act(async () => {
      await component.root.findByProps({ id: 'options-columns-id' }).props.onChange(false);
    });

    expect(save).toHaveBeenCalledWith('columns', ['class']);

    await act(async () => {
      await component.root.findByProps({ id: 'options-columns-licence' }).props.onChange(true);
    });

    expect(save).toHaveBeenCalledWith('columns', ['id', 'class', 'licence']);
  });
});
