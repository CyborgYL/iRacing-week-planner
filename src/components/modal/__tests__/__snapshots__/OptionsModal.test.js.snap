// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/modal/OptionsModal renders closed 1`] = `null`;

exports[`components/modal/OptionsModal renders with checkbox saving 1`] = `
<div
  className="ReactModal__Overlay"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 0.75)",
      "bottom": 0,
      "left": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
    }
  }
>
  <div
    aria-modal={true}
    className="ReactModal__Content Modal__Bootstrap modal-dialog modal-lg"
    id="optionsModal"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={Object {}}
    tabIndex="-1"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <button
          aria-label="Close"
          className="close"
          onClick={[Function]}
          type="button"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
        </button>
        <h4
          className="modal-title"
        >
          Options
        </h4>
      </div>
      <div
        className="modal-body"
        style={
          Object {
            "maxHeight": "55vh",
            "overflowY": "auto",
          }
        }
      >
        <div
          className="container-fluid"
        >
          <h5>
            Columns
          </h5>
          <div
            className="row"
          >
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-id"
                >
                  <input
                    checked={true}
                    disabled={false}
                    id="options-columns-id"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  ID
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-class"
                >
                  <input
                    checked={true}
                    disabled={false}
                    id="options-columns-class"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Class
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-licence"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="options-columns-licence"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Licence
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-type"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="options-columns-type"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Type
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox disabled"
              >
                <label
                  htmlFor="options-columns-series"
                >
                  <input
                    checked={false}
                    disabled={true}
                    id="options-columns-series"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Series
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-track"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="options-columns-track"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Track
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-car"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="options-columns-car"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Car
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-start"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="options-columns-start"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Start
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-end"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="options-columns-end"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  End
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-racelength"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="options-columns-racelength"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Length
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-official"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="options-columns-official"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Official
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-fixed"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="options-columns-fixed"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Fixed
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-raceTimes"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="options-columns-raceTimes"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Race times
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-nextRace"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="options-columns-nextRace"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Next race
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-seriesEnd"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="options-columns-seriesEnd"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Season end
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-rain"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="options-columns-rain"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Rain
                </label>
              </div>
            </div>
            <div
              className="col-xs-3 col-md-2"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="options-columns-seriesLink"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="options-columns-seriesLink"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Link
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="modal-footer"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;
